Fetch API를 이용한 데이터 처리

sangwon21 opened the pull request:
- what
  - 데이터를 비동기적으로 불러오는 작업을 진행했습니다.
  - Promise, async와 await을 활용하여 비동기를 제어했습니다.
  - webpack을 활용하여 js파일과 css 파일을 번들링했습니다.
  - 서버 환경(express)을 구축했습니다.

- how
  - fetch API를 통해, 서버에 데이터를 요청한 이후, 받아온 데이터를 다뤘습니다.
  - 받아온 데이터를 localStorage에 캐싱했습니다. 
  - webpack.config.js 파일을 통해, webpack을 번들링하는 과정에서 필요한 설정을 미리 정했습니다.
  - express-generator를 활용해, 미리 정해진 규격대로 express서버를 구축했습니다. 

- why
  - fetch API를 통해 비동기적으로 데이터 요청을 처리할 수 있었습니다. 
  - webpack을 통해 css와 js를 하나의 component로 묶울 수 있어서 UI component 의미가 강화되었습니다.

- 더 해야될 일
  - promise와 async, await을 조금 더 이해하기
  - 슬라이더 동작 개선

Commit:
[Reese, won] Slider UI HTML, CSS 작성 (#114)

* Complete basic structure

- Add util.js for the sake of convenience

* Complete header html rendering

* Complete card html rendering

* Complete button html rendering

* Complete header css

* Complete card css

* Complete button css

* Format Style

- Consistency in tab indent

* Fix type

sangwon21 committed:
feat: Header CSS 수정

why: 아마존 슬라이더와 비슷하게 CSS를 수정했다. 기존의 문구만 있던
부분에서, 배경 이미지를 준 CSS로 수정했다.

sangwon21 committed:
feat: 헤더 클릭 시 이벤트 추가

why: 헤더의 그림을 클릭하면, 해당 상자가 커져야 한다. 해당 상자 옆으로
그림자 효과도 추가되어야 한다. 마지막으로 해당 상자의 숨겨진 점들이
드러나야 한다. 이를 위해서 이벤트를 추가했는데, 이벤트 객체를 분리해서
온전히 이벤트 추가만 담당하게 만들었다. 하지만, 이벤트 버블링을 고려한
함수를 만들다 보니, 함수가 하는 일이 지나치게 커졌다. 이를 줄이기 위한
노력이 필요하다.

sangwon21 committed:
feat: 가짜 데이터 변경

why: 데이터 구조에 맞게 가짜 데어터를 변경했다. 프로그램에서 필요한
데이터를 추가했다. 이에 맞춰, 데이터를 다루기 편하게, 객체 형태로
저장했다.

sangwon21 committed:
feat: 이벤트 객체 생성 및 분리

why: 기존코드에서 slider-event에서 모든 이벤트의 책임을 지고 있었다.
하지만 헤더 영역과 슬라이더 영역에서 서로 필요한 이벤트의 성격이 다름에
따라 분리할 필요성이 생겼다. 코드를 작성하다 보니, 이벤트를 실제
이벤트가 적용되는 영역에 붙여주는 코드가 중복되는 것을 발견했다. 이
부분은 상속을 통해 다른 하나의 객체로 분리해 주었다.

sangwon21 committed:
feat: 슬라이더 기능 구현

why:
- 버튼의 next나 prev를 클릭하면 카드가 다음 방향이나 이전 방향으로 움직이는 기능을 구현했다. 개발할 때, 편의상 overflow: hidden값을 주지 않았다. 카드 1장당 id를 1,2,3... 이런 식으로 정수로 주고, css를 이런정수에 적용시켜 보려고 했다. 하지만 css 파일에서 정수를 인식하지 못하여, 대신 card + 정수 이런 id값을 활용하였다.
- 슬라이더 구현 방식은 다음과 같다. 카메라는 긴 띠의 중앙만 항상 촬영한다. 그래서 미리 띠를 중앙으로 옮겨야한다. 이 때, 옮기는 방식은 다양한데, transform만큼은 활용하면 안 된다. 왜냐하면 후에 상술할, 카드 이동방식이랑 css가 겹쳐, 원하는 효과가 나타나지 않는다. 여기서는, left에-값을 주어, 중앙에 배치되게끔 했다. 사용자가 버튼을 클릭하면, 해당 방향에 맞게 띠를 한 칸 이동해준다. transition이 끝나면 옮긴 칸이 중앙에 오게끔 배열을 조작한다. 이 때, transition 효과를 none으로 주어, 어떠한 움직임도 화면에 나타나지 않게한다. 배열 조작 후, translateX(0)으로 주어 제자리에 있게끔 한다.

sangwon21 committed:
feat: 헤더의 점 클릭시 선택 표시

why: 헤더를 클릭했을 때, 점을 표시해준다. 이벤트가 다양한 영역에서
발생할 수 있기 때문에, if문을 통하여 각각의 경우를 처리했다. 또한, dot을
정확히 클릭했을 때, dot이 사라지는 경우가 발생했다. 이는 invisible
클래스를 잘못된 tag에 줘서 발생항 문제로, 다시 제대로 된 tag에 줬다.

sangwon21 committed:
feat: index 기반 슬라이더 기능 구현

why: 헤더 버튼을 index로 이용하여 슬라이더를 구현할 예정이다. 이를 위해,
기존의 버튼을 클릭하면 1칸씩 이동하던 방법을 index기반으로 이동하는
방식으로 교체하였다.

sangwon21 committed:
feat: 동그라미를 클릭시 슬라이더 이동 구현

why: 동그라미를 클릭하면 이동하는 이벤트를 구현했다. 인덱스 기반에서
거리 기반으로 이동하게끔 함수를 바꿨다. 거리 구하는 로직이 복잡해지는
만큼, 나머지 로직이 편해졌다.

sangwon21 committed:
chore: slider.html 파일 이름 변경

why: git page 등록 편의성을 위하여 slider.html을 index.html로 변경했다.

sangwon21 committed:
feat: 클릭시 카드 이동

why: 클릭시 그 카테고리의 첫번째 카드로 이동하는 기능을 구현했다. 이를
위해, 객체들 간 참조가 필요하다고 생각했다. 하지만 지금 구조에서 다른
이벤트 객체를 참조하지는 못한다. 그래서 구조를 바꾸는 고민도 했지만,
답은 의외로 click()이라는 함수를 통해 클릭을 흉내내는 방법으로 해결했다.
click을 하면 이벤트가 발생하니, click을 통하여 나머지 이벤트들을
해결했다.

sangwon21 committed:
chore: index.html 파일 이름 변경

why: slider.html 이름이 더 적절하다.

sangwon21 committed:
chore: update README

why: README에 동작하는 페이지 주소를 올렸다.

sangwon21 committed:
refact: 특정 상수 값 변수화

why: 알고리즘 작성 중에 필요한 값들이 기존에는 상수로 들어가 있었는데
변수화를 진행함으로써, 좀 더 알고리즘에 대한 이해를 높일 수 있게끔
작성했다.

sangwon21 committed:
refact: 공통된 부분 함수로 분리

why: 카드 이동 후, 나타난 카드에 효과를 주는 부분을 css클래스로
작성했다. 이 클래스를 줌으로써, 효과를 준다. 이 부분이 여러 함수에서
활용되기에, 독립적인 함수로 분리했다.

sangwon21 committed:
refact: 초기화 함수 부모 객체로 분리

why: event 객체에서 초기화 과정에 필요한 함수가 겹쳤다. 따라서 필요한 이
공통 함수를 부모 객체에 매서드로 주었다.

sangwon21 committed:
feat: 첫 화면 제공 시 필요한 CSS 클래스 추가

why: 첫 화면이 띄워졌을 때, 필요한 CSS 클래스들을 추가했다. 페이지에서
제공되는 효과들은 CSS 클래스에 기반하고 있는데, 첫 화면 렌더링 시 이
클래스들을 tag에 추가하지 못해, 첫 화면에서는 효과가 없었다. 이를
해결하기 위하여, 첫 번째 렌더링 때만 해당 CSS 클래스들을 추가했다.

sangwon21 committed:
chore: update README

why: git page 주소가 수정되어 README에 업데이트 했다.

sangwon21 committed:
refact: 공통된 부분 함수로 분리

why: 이벤트 콜백 함수에서 SliderEvent 클래스에 필요한 데이터를
초기화하는데 반복되는 코드가 있었다. 반복적으로 작성되는 코드를 분리하여
하나의 독립적인 함수로 만들었다.

sangwon21 committed:
refact: 반복되는 코드나 의미가 불명확한 부분을 함수로 분리

why: 반복적으로 사용되는 코드를 객체의 멤버 변수를 활용하여 없앴다.
특히, 반복되는 코드가 계속해서 dom 트리를 탐색하는 작업이기에, 성능에
도움이 된다. 또한, getDistance에서 알고리즘을 담당하는 부분에서, 의미가
불명확한 부분이 있기에, 작은 함수로 나누어 의미를 명확한 의미를
부여했다.

sangwon21 committed:
feat: express로 서버 구축

why: local 서버에서 json데이터를 받아오기 위하여 express로 서버를
구축했다.

sangwon21 committed:
feat: json 데이터 전송 API 작성

why: json 데이터를 요청했을 때, 응답할 수 있는 API를 서버에 만들었다.
GET 방식으로 json 데이터를 요청하면, 해당 API는 미리 정적(static)하게
존재하던 json데이터 파일을 읽어온 다음, json 데이터를 보낸다. 이 때,
response의 header설정을 "Access-Control-Allow-Origin"으로 설정했는데,
이는 CORS 문제에 대응하기 위함이다.

sangwon21 committed:
feat: client 파일로 분리 및 module 적용

why: server 폴더 생성으로 client 폴더의 필요성이 생겼다. 그래서
client 관련 파일들을 client 폴더에 넣기로 결정했다. 이렇게 폴더를
구획으로 나누면, server와 client의 구별이 생긴다. 더불어 module도
적용했다. 기존에는 html에서 필요한 모든 script들을 다 불러오는 방식으로
작업을 했다. 이는 전역적으로 변수들이 쓰여 문제를 야기할 위험성이
있었다. 그래서 import, export 방식을 활용한 module 시스템을 사용하기로
결정했다.

sangwon21 committed:
feat: 프로그램 내 상수용 객체 추가

why: 프로그램 도중에 활용되는 각종 상수들을 한 곳에 모아 정리하기
위하여, 객체를 만들어 관리하기로 결정했다.

sangwon21 committed:
feat: data fetch 부분 추가

why: data를 비동기적으로 받아온다. 받아온 데이터를 이용하여 Slider를
생성하는 '동기적' 작업을 진행해야 한다. 비동기적 작업을 동기적으로
보이게 할 수 있는 await, async를 활용했다.

sangwon21 committed:
feat: 서버에서 받아온 데이터 cache 작업 추가

why: 서버에서 받아온 card data를 캐싱한다. 기존에 static함수로 card
data를 서버에서 fetch해왔는데, 캐싱작업을 하다 보니, static으로 함수를
작성하니, 캐싱하는 부분이 많이 복잡해졌다. 의미로도 당위성이 떨어지는
static함수를 삭제하고 모두 class 메서드로 선언하고, 캐싱할 변수를 class
변수로 선언했다.

sangwon21 committed:
feat: webpack 적용

why: 간결하게 파일 관리를 하고 싶어서, webpack을 사용했다. 추후에 더
공부가 필요하다. 일단, 하나의 js파일에서 필요한 js파일들을 모으고, 또한
해당 component에 필요한 css파일까지 번들링하는 작업까지 끝마쳤다.

sangwon21 committed:
feat: localStorage를 이용한 캐싱

why: 클래스의 멤버 변수로 가지고 있는 것보다 브라우저의 localStorage를
통해 캐싱을 하는 방법이 조금 더 적합하다고 생각하여 바꿨다.

sangwon21 committed:
fix: 충돌 해결

sangwon21 committed:
Merge branch 'sangwon21' into fetch

crongro commented:
커멋단위, 커밋로그 좋아요 💯

crongro commented:
dist아래에  빌드된 파일은 소스코드에 올리지 않긴해요~

crongro commented on client/src/button.js:
```diff
@@ -0,0 +1,19 @@
+import css from "./button.css";
+
+class Button {
```
디렉토리 이름을
client/src/view/button.js 라고 하면 더 좋을 듯.

sangwon21 replied:
의미가 더 살아나는 방향인 것 같습니다. view로 한 번 묶어보겠습니다.

crongro commented on client/src/card.css:
```diff
@@ -0,0 +1,173 @@
+    overflow: hidden;   
+}
+
+.card-wrapper {
```
sass 익숙한가요? 써보세요.

sangwon21 replied:
sass 한 번도 써본 적이 없지만 해보겠습니다

crongro commented on client/src/card.js:
```diff
@@ -0,0 +1,27 @@
+class Card {
+    constructor(cardData) {
+        this.cardData = cardData;
+        this.idIndex = 1;
```
id보다는 cardIndex 라고 해도 될 듯. (카드 맞으면)

sangwon21 replied:
cardIndex가 id보다 조금 더 명확한 의미를 가지는 것 같습니다.

crongro commented on client/src/card.js:
```diff
@@ -0,0 +1,27 @@
+    sortByCategory(category) {
+        return this.cardData
+            .filter(data => data.category === category)
+            .reduce((result, data) => (result += `<li class="card ${this.idIndex === 1 ? "selected-card" : ""}" id=card-${this.idIndex++}><div class="card-copy"><span class="card-category ${data.category}">${data.cardCategory}</span><div class="card-headline">${data.cardHeadline}</div><div class="card-body">${data.cardBody}</div><a class="card-cta">${data.cardCta}</a></div></li>`), "")
```
HTML문자열이 가로로 길어서 보기 어렵네요. 개행해도 잘 될거에요

sangwon21 replied:
협업하는 다른 사람의 시선을 염두에 두고 코딩을 하도록 노력하겠습니다.

crongro commented on client/src/card.js:
```diff
@@ -0,0 +1,27 @@
+        const streamList = this.sortByCategory("stream");
+        const shopList = this.sortByCategory("shop");
+        const readList = this.sortByCategory("read");
+        const moreList = this.sortByCategory("more");
```
this.sortByCategories(["ship", "stream", "shop", "read", "more"]);
이렇게 동작해서 중복코드를 줄이면 좋을 듯.

sangwon21 replied:
코드를 작성하면서 찝찝한 부분이었는데, 좋은 해결 방향성을 제시해줘서 감사합니다.

crongro commented on client/src/constants.js:
```diff
@@ -0,0 +1,3 @@
+export const URL = {
+    SERVER: "http://localhost:8080/cards"
```
development 와 production 을 고려해야 할텐데요.

URL = {
 API_SERVER : {
     DEV : "",
     STAGE : "",
  }
}

이런식으로 여러가지 개발환경을 염두해서 만들면 더 좋을 듯.

sangwon21 replied:
질문이 있습니다. node 환경에서는 process.env를 통해 production환경과 development환경을 구별할 수 있었던 것 같은데, 브라우저 환경에서는 따로 구별하는 방법이 있을까요?

crongro commented on client/src/event-manager.js:
```diff
@@ -0,0 +1,24 @@
+import HeaderEvent from "./header-event.js";
+import SliderEvent from "./slider-event.js";
+import { $, $$ } from "./util.js";
```
as 앨리어스를 사용해보세요.

crongro commented on client/src/event-manager.js:
```diff
@@ -0,0 +1,24 @@
+        $$(".dot").forEach(dot => {
+            this.sliderEvent.addEvent(dot, "click", this.sliderEvent.dotEventListener.bind(this.sliderEvent));
+        })
+        this.sliderEvent.addEvent($(".button2"), "click", this.sliderEvent.nextButtonListener.bind(this.sliderEvent));
```
arrow function을 사용하면 bind를 쓰지 않아도 될 수 있다는거 알고있죠?
왜 그런지 설명할 수 있어야 함.

crongro commented on client/src/header-event.js:
```diff
@@ -0,0 +1,66 @@
+import MyEvent from "./my-event.js";
+
+class HeaderEvent extends MyEvent {
```
MyEvent라기 보다는 CoreEvent, CommonEvent, ParentEvent, EventManager 등의 이름이 더 나을 듯.

crongro commented on client/src/header-event.js:
```diff
@@ -0,0 +1,66 @@
+    handleHeaderPart(target) {
+        target.parentNode.classList.add('header-selected');
+        target.parentNode.querySelector(".circles").classList.remove("invisible");
+        target.parentNode.querySelector(".dot").classList.add("selected-dot");
```
const dotNode =  target.parentNode.querySelector(".dot")

재사용하기!

crongro commented on client/src/header-event.js:
```diff
@@ -0,0 +1,66 @@
+        this.initializeStatus();
+
+
+        if (target.classList.contains("header_title") === true) {
```
여긴 개행이 2줄도 있고 너무 넉넉하네요 ㅎ

crongro commented on client/src/header.js:
```diff
@@ -0,0 +1,30 @@
+
+    returnCardLengthDivString(length) {
+        let result = "";
+        for (let i = 0; i < length; i++) {
```
가급적 forEach 사용~

sangwon21 replied:
특정 숫자만큼 돌고 싶을 때, for문을 어떻게 대체할 수 있을까요?

crongro replied:
아 가능하지만 그냥 for문이 더 낫겠네요!

crongro commented on client/src/main.js:
```diff
@@ -0,0 +1,39 @@
+
+
+const addEvents = () => {
+    const eventManager = new EventManager();
```
여기 eventManager가 있군요.

crongro commented on client/src/main.js:
```diff
@@ -0,0 +1,39 @@
+    eventManager.addEvents();
+}
+
+const constructHTML = async () => {
```
생성자의 단어와 비슷해서.. 전 별로

sangwon21 replied:
renderHTML로 교체(?)하겠습니다.

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+        this.distance = 0;
+        this.previousIndex = 0;
+        this.currentIndex = 0;
+        this.listLength = 17;
```
default 값이라는 걸 의미하면 더 좋을 듯. 

this.defaultOption = {
        this.currentIndex :0,
        this.listLength : 17
}

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+        if (this.isRightWay(distance) === true) {
+            this.isPrevious = false;
+            cardWrapper.style.transition = "transform 1s";
+            cardWrapper.style.transform = `translateX(${-1080 * distance}px)`;
```
1080 magic number! 변수화라도 시키세요.

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+        this.setCurrentIndexAsPreviousIndex()
+        this.currentIndex = this.getPreviousIndex(this.previousIndex);
+        this.distance = -1;
+        this.moveByDistance(-1);
```
혹시 ,
 this.moveByDistance(this.distance) 가 아닐까요?

sangwon21 replied:
네, 말씀하신 방법대로 하면 의미가 명확해져서 그렇게 바꾸겠습니다.

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+    transitionEndEvent(event) {
+        const cardWrapper = $(".card-wrapper");
+        cardWrapper.style.transition = 'none';
+        const childNodes = Array.from(cardWrapper.children);
```
spread opeartor 도 활용해보세요. from말고요.

sangwon21 replied:
spread operator가 유사 배열에도 적용이 가능하군요.  [참고 자료](https://stackoverflow.com/questions/222841/most-efficient-way-to-convert-an-htmlcollection-to-an-array)

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+            this.distance = (target - middle);
+            return;
+        }
+        if (1 <= target && target <= middle + this.halfListLength - this.listLength) {
```
if문 안에 내용이 너무 길고 복잡해요.
변수로 저장을 먼저 하는건 어때요.

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+        return;
+    }
+
+    getDistance(middle, target) {
```
getDistance 메서드지만 거리를 반환하지는 않네요.

sangwon21 replied:
get 메서드와 set메서드를 햇갈렸네요. setDistance로 바꾸겠습니다.

crongro commented on client/src/slider-event.js:
```diff
@@ -0,0 +1,149 @@
+        this.getDistanceInNormalOrder(middle, target);
+    }
+
+    dotEventListener(event) {
```
dot 이 무엇의미인지 전 잘 모르겠네요.

sangwon21 replied:
dot은 헤더(상단의 작은 박스) 속 동그라미를 의미합니다. 좀 더 명징한 이름이 있을까요?

crongro replied:
동그라미도 뭔가 의미가 있지 않나요? 
형태를 표현하지 말고..의미를 부여하는거 가능하면 그렇게 해보시죠.

crongro commented on client/src/main.js:
```diff
@@ -0,0 +1,39 @@
+    const slider = new Slider();
+    let cards = window.localStorage.getItem('cards');
+    if (!cards) {
+        const cardData = await slider.fetchCardData();
```
fetch 사용과 promise 반환을 잘 이해하신 듯.

crongro reviewed (changes requested):
수고했네요.

리뷰 남긴 내용 수정해보세요~

sangwon21 committed:
fix: 애니메이션 중 이벤트 발생 오류

why: 애니메이션 중 다른 이벤트가 발생하여 이동 애니메이션이 완벽하게
동작하지 못하는 오류가 있었다. 이를 막기 위해, 이동 애니메이션 중에는
다른 이벤트가 동작하지 않도록 코드를 수정하였다.

sangwon21 committed:
fix: 같은 박스를 클릭했을 때 오류 해결

why: 같은 박스를 두 번 클릭하면, 이 이후부터, 다음 동작이 존재하지 않는
오류를 해결했다. 이를 해결하기  위해, 하나의 상태 객체를 공유하는 방법을 썼다.
상태를 공유하는 객체를 이벤트 객체들에서 활용하여, 서로 상태를 공유할 수
있게 되었다.

sangwon21 committed:
refactor: client 폴더 구조 개편(명확한 역할 부여)

why: client 폴더에 view, event, util 라는 폴더를 만들었다. 폴더를
나눔으로써, 파일 간 역할이 명확히 나눠졌다. 또한 명확한 역할이 부여되어,
코드를 읽기가 조금 더 편해졌다.

sangwon21 committed:
fix: 폴더 생성으로 인한 파일 경로 수정 및 의미 없는 변수명 변경

why: 폴더 구조 생성으로 인하여 파일 경로를 수정해야 했다. 또한, id 보다
조금 더 해당 변수의 의미를 살릴 수 있는 cardIndex로 변수명을 변경했다.

sangwon21 committed:
fix: HTML 템플릿 개행을 통한 가독성 향상

why: HTML 템플릿을 활용할 때, 길게 HTML을 늘어놓았다. 코드의 가독성을
해쳐서, HTML 구조처럼 템플릿을 구성했다.

sangwon21 committed:
refactor: 함수 파이프라인 설계를 통한 가독성 개선

why: 기존에 5개의 카테고리에 대하여 한 번씩 모두 함수를 호출하고 있었다.
이를 개선하기 위하여 함수를 한 번만 호출하는 대신, 그 함수의 인자로
배열을 받기로 결정했다. 함수 안에서, 클로져를 만들어, 내부 함수를,
배열을 한 바퀴 돌면서 실행하는 방식으로 변경했다.

sangwon21 committed:
refactor: get 메서드를 set 메서드로 교체

why: getDistance를 setDistance로 이름을 교체했다. 왜냐하면, 이 함수를
통해 변경되는 값은 객체의 멤버 변수이기 때문이다. 그렇기 때문에, set
함수명이 적절하다.

sangwon21 committed:
refactor: 개발 환경과 프로덕션 환경을 구별하는 서버 URL 설정

why: 개발 과정에서 요청을 보내야 하는 주소와 실제 제품에서 요청을 보내야
하는 주소가 달라야 한다. 그래서, 이러한 구분을 하기 위해서 주소 객체에
각각의 상황에 맞는 주소 객체를 뒀다.

sangwon21 committed:
fix: 일반적인 MyEvent 이름 대체

why: 너무 범용적인 MyEvent이라는 이름 대신, CoreEvent이라는 이름을
활용하기로 결정했다. MyEvent이라는 이름 보다 CoreEvent이라는 이름이 훨씬
더 클래스가 가지고 있는 의도를 잘 드러낸다.

sangwon21 committed:
fix: 유사 배열에 spread operator 활용

why: Array.from이라는 함수 대신 spread operator를 활용했다. 유사 배열에
spread operator를 활용할 수 있다는 걸 처음 알게 되었다. 코드의 일관성을
위해, Array.from 대신 spread operator를 활용했다.

sangwon21 committed:
fix: constructHTML 함수명 변경

why: constructHTML 함수명이 생성자 함수와 비슷해보이는 부분이 있어서
renderHTML로 함수명을 변경했다.

sangwon21 committed:
refactor: SliderEvent 객체 생성자에 default 옵션 부여

why: SliderEvent 객체에 옵션을 줌으로써, 옵션에 따라 다르게
동작할 수 있도록 바꿨다. 또한 미리 값을 부여하여, 옵션이 없으면, default
옵션값이 사용되게끔 구현했다.

sangwon21 committed:
fix: 카드 길이 변수화 및 default option에 추가

why: 카드 길이를 변수화해서 default option에 추가했다. 카드 길이도
사용자가 설정할 수 있게끔 해서, 조금 더 유연한 객체가 되게끔 했다.

sangwon21 committed:
refactor: nextButtonListener과 previousButtonListener 상수로 써져 있던 부분을 변수로 교체

why: 상수로 써져 있던 부분을 this.distance로 교체했다. 이렇게 함으로써,
조금 더 의미가 명징해지는 효과를 얻을 수 있다.

sangwon21 committed:
format: 포맷팅

sangwon21 committed:
refactor: HeaderEvent에서 DOM 캐싱을 통한 성능 개선

why: HeaderEvent에서 DOM을 다룰 때, 계속해서 querySelector를 통해 노드를
계속 탐색했다. 하지만 변수에 '캐싱'하면 한 번 갔던 노드를 또 탐색하지
않아도 되니, 성능의 향상을 기대할 수 있다.

sangwon21 committed:
fix: dotEventListener 이름에 조금 더 의미 부여

why: dotEventListener라는 이름이 명확하지 못해, 조금 더 의미를 부여해줄
수 있는 smallBoxIndexEventListener로 함수 이름을 바꿨다.

sangwon21 committed:
fix: getDistance 알고리즘 비교 구문에 변수 부여를 통한 구체적 의미 부여

why: getDistance 알고리즘에서 if문 속 조건문이 명징하지 못해, 적절한
변수를 두어(distanceFromMiddle) 의미 부여를 했다.

sangwon21 committed:
refactor: $$에 이름 부여

why: $$ 대신 SelectAll를 써봤다. export한 이름 그대로를 활용하지 않고
다른 이름으로 재정의한 이름을 쓸 수 있다.

sangwon21 commented:
## 고치거나 바꾼 부분
- 변수명이나 함수명 혹은 클래스명에 범용적인 이름 대신 역할에 적확한 이름을 사용해 보려고 노력했습니다.
- 슬라이더 동작을 개선했습니다. 애니메이션 이동 거리 알고리즘이 리팩토링 과정에서 문제가 발생하여 해결했습니다. 더불어 애니메이션 진행 중에 또 다시 애니메이션 이벤트가 발생하면 카드 순서가 헝클어지는 문제가 발생했는데, 애니메이션 진행 중에는 또다른 애니메이션이 발생하지 않는 방향으로 수정했습니다.
- DOM 캐싱을 통하여 불필요한 DOM 노드 탐색을 줄였습니다.
- 함수 커링을 통하여 한층 더 가독성이 좋은 코드를 작성했습니다.
- css 대신 scss를 활용했습니다.

## 더 해볼 부분
- webpack에 유용한 loader 및 plugin 사용해보기
-  함수 커링에 대하여 공부해보기
- 서로 다른 모드(production, development 등)에서 다르게 동작하게 연구해보기

[데모](https://sangwon21.github.io/amazon-carousel/client/slider.html)

sangwon21 committed:
refact: css에서 scss로 전환

why: 기존 css에서 scss로 전환하면서 css를 프로그래밍 하듯이 다룰 수 있게
되었다. 예를 들어, 의미 없는 상수 px를 변수에 대입함으로써, 변수에
의미를 부여하게 되었다. 이를 작성하는 사람도, 읽는 사람도 의미를 명확히
전달하고, 받을 수 있게 되었다.

sangwon21 committed:
fix: 트랜지션 속도 향상, HTML 템플릿 가독성 향상, 함수명 변경

why: 트랜지션 속도를 향상시킴으로써, 조금 더 부드러운 애니메이션을
보여줄 수 있게 되었다. HTML 템플릿을 구조화하고, 함수명을 해당 역할에
맞게끔 변경하여 가독성을 향상시켰다.

crongro commented on client/src/view/header.scss:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  &:nth-child(5) {
+    background-image: url(https://web.archive.org/web/20180717213726im_/https://images-na.ssl-images-amazon.com/images/G/01/marketing/prime/detail_page/InteractivePlane_TILES_SpriteSheet_E._CB513285352_.png),
```
참고로 콘텐츠의 일부라면, img태그를 사용해야 한다는 점도 기억해주세요.

crongro commented on client/src/view/header.scss:
```diff
@@ -0,0 +1,89 @@
+}
+
+.selected-dot {
+  background-color: #ffffff;
```
축약표기법을 사용해보시죠.. 
#fff

crongro reviewed (approved):
네 수정사항 확인했어요.
css preprocessor사용시도도 좋네요. 다양한 기능이 많은거 같으니  하나 둘 씩 기능을 사용해보세요.
